math/optimized_helper_functions.h

```diff
@@ -0,0 +1,100 @@
+/*
+	Copyright 2021 Kenn Sebesta
+
+	This file is part of the VESC firmware.
+
+	The VESC firmware is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    The VESC firmware is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+    */
+
+#ifndef OPTIMIZED_HELPER_FUNCTIONS_H_
+#define OPTIMIZED_HELPER_FUNCTIONS_H_
+
+#include <stdbool.h>
+#include <stdint.h>
+#include "datatypes.h"
+#include <math.h>
+#include <stdlib.h>
+
+// This takes advantage of the IEEE 754 standard, which requires that the sign
+// bit of single-precision floats is the last bit.
+// It is faster than `signbit()` because it does not have a bit shift and does no type checking.
+// On the flip side, it is dangerous because it does no checking of any kind. It will
+// not care if the input is not a `float`, if it is `NaN`, etc...
+// It is guaranteed to work on all systems which adhere to IEEE 754.
+// This works on both big- and little-endian systems.
+// The macro returns 1 if a `float` is negative, 0 if otherwise
+#define IS_FLOAT_IEEE754_NEGATIVE(x) (*(uint32_t *)(&(x)) & (1U<<31))
+
+
+/**
+ * @brief calculate_mod_alpha_beta_filter_sign Exploits the fact that two of the three
+ * currents will always have the same sign, and the third will always have the opposite sign.
+ * @param mod_alpha_filter_sgn Returns a value on the set of [-4/3, -2/3, 2/3, 4/3]
+ * @param mod_beta_filter_sgn Returns a value on the set of [-2/sqrt(3), 0, 2/sqrt(3)]
+ * @param ia_filter The filtered i_a
+ * @param ib_filter The filtered i_b
+ * @param ic_filter The filtered i_c
+ */
+inline __attribute__((always_inline)) void calculate_mod_alpha_beta_filter_sign(float *mod_alpha_filter_sgn, float *mod_beta_filter_sgn, float ia_filter, float ib_filter, float ic_filter) {
```

Comment:
Just maybe:
What about collecting sign bits into 3bit integer and use it to index 8 element table with mod_alpha_filter_sgn / mod_beta_filter_sgn values?

Reply:
So something like:

```
#define SIGN_BIT_FLOAT_IEEE754(x) (*(uint32_t *)(&(x)) & (1U<<31))

// Possible output values of the truth table
const float A[6] = {0, 1, 3, 4, 5, 6};  // NOTE: Not correctly populated yet
const float B[6] = {0, 1, 3, 4, 5, 6};

void calculate_mod_alpha_beta_filter_sign(float *mod_alpha_filter_sgn, float *mod_beta_filter_sgn, float ia_filter, float ib_filter, float ic_filter) {

    // Create a truth table index in 0b000..0b111.
    int idx = (SIGN_BIT_FLOAT_IEEE754(ia_filter) >> 29) | (SIGN_BIT_FLOAT_IEEE754(ia_filter) >> 30) | (SIGN_BIT_FLOAT_IEEE754(ia_filter) >> 31);

    *mod_alpha_filter_sgn = A[idx];
    *mod_beta_filter_sgn = B[idx];
}
```

That's beautiful in its simplicity, and has the nice effect of being constant in execution time. I'll program it up for real and test it.

Reply:
Personally, I would use 
```
const struct { float a,b; } tab[8] = {{1,2},...};
```
Also it is good idea to specify all 8 possibilities, ideally with 'safe' value for 'impossible' input combination ... 

As for generated code - ARM is well optimized for bitfield access, so shift operations are nearly free. And avoiding branch saves a cycle.